PWGCF/GenericFramework/Tasks/flowGenericFramework.cxx

```diff
@@ -278,163 +320,188 @@ struct GenericFramework {
   {
     if (tracks.size() < 1)
       return;
     if (centrality < centbinning.front() || centrality > centbinning.back())
       return;
     float vtxz = collision.posZ();
-
-    if (cfgFillQA) {
-      QAregistry.fill(HIST("cent_nch"), tracks.size(), centrality);
-      QAregistry.fill(HIST("vtxZ"), vtxz);
-    }
+    if (cfgFillQA)
+      (dt == kGen) ? QAregistry.fill(HIST("vtxZ_gen"), vtxz) : QAregistry.fill(HIST("vtxZ"), vtxz);
     fGFW->Clear();
     float l_Random = fRndm->Rndm();
-    float weff = 1, wacc = 1;
-
     for (auto& track : tracks) {
-      if (cfgFillWeights)
-        fWeights->Fill(track.phi(), track.eta(), vtxz, track.pt(), centrality, 0);
-      if (cfg.mEfficiency)
-        weff = cfg.mEfficiency->GetBinContent(cfg.mEfficiency->FindBin(track.pt()));
-      else
-        weff = 1.0;
-      if (weff == 0)
-        continue;
-      weff = 1. / weff;
-      if (cfg.mAcceptance)
-        wacc = cfg.mAcceptance->GetNUA(track.phi(), track.eta(), vtxz);
-      else
-        wacc = 1;
-      registry.fill(HIST("phi_eta_vtxZ_corrected"), track.phi(), track.eta(), vtxz, wacc);
-      bool WithinPtPOI = (ptpoilow < track.pt()) && (track.pt() < ptpoiup); // within POI pT range
-      bool WithinPtRef = (ptreflow < track.pt()) && (track.pt() < ptrefup); // within RF pT range
-      if (WithinPtRef)
-        fGFW->Fill(track.eta(), fPtAxis->FindBin(track.pt() - 1), track.phi(), wacc * weff, 1);
-      if (WithinPtPOI)
-        fGFW->Fill(track.eta(), fPtAxis->FindBin(track.pt()) - 1, track.phi(), wacc * weff, 2);
-      if (WithinPtPOI && WithinPtRef)
-        fGFW->Fill(track.eta(), fPtAxis->FindBin(track.pt()) - 1, track.phi(), wacc * weff, 4);
+      ProcessTrack(track, centrality, vtxz);
+    }
+    for (uint l_ind = 0; l_ind < corrconfigs.size(); l_ind++) {
+      FillFC((dt == kGen) ? fFC_gen : fFC, corrconfigs.at(l_ind), (cfgUseNch) ? tracks.size() : centrality, l_Random);
+    }
+  }
 
+  template <typename TrackObject>
+  inline void ProcessTrack(TrackObject const& track, const float& centrality, const float& vtxz)
+  {
+    float weff = 1, wacc = 1;
+    float pt, eta, phi;
+    if constexpr (framework::has_type_v<aod::mctracklabel::McParticleId, typename TrackObject::all_columns>) {
+      if (track.mcParticleId() < 0 || !(track.has_mcParticle()))
+        return;
+      auto mcParticle = track.mcParticle();
+      pt = mcParticle.pt();
+      eta = mcParticle.eta();
+      phi = mcParticle.phi();
+      if (!mcParticle.isPhysicalPrimary() || eta < etalow || eta > etaup || pt < ptlow || pt > ptup)
+        return;
+      if (cfgFillWeights)
+        fWeights->Fill(phi, eta, vtxz, pt, centrality, 0);
+      if (!setCurrentParticleWeights(weff, wacc, phi, eta, pt, vtxz))
+        return;
+      registry.fill(HIST("phi_eta_vtxZ_corrected"), phi, eta, vtxz, wacc);
+      QAregistry.fill(HIST("phi"), phi);
+      QAregistry.fill(HIST("eta"), eta);
+      QAregistry.fill(HIST("pt"), pt);
```

Comment:
For the next iteration consider a `FillQA` templated function with the data type and an abstract track/particle object
Then you will not repeat the fills and if in the future your QA needs increase you will need to touch only in one place